Python + Aspire Dashboard docs (#1561)
dotnet/docs-aspire: docs/fundamentals/dashboard/standalone-for-python.md

---
title: Using Aspire Dashboard with Python Applications
description: How to use the Aspire Dashboard in a Python application.
ms.date: 08/27/2024
ms.topic: tutorial
ms.author: aapowell
---

# Tutorial: Using the Aspire Dashboard with Python Applications

The [Aspire dashboard](overview.md) provides a great UI for viewing telemetry, and is available as a standalone container image that can be used with any OpenTelemetry-enabled app. In this article, you'll learn how to:

> [!div class="checklist"]
>
> - Start the Aspire dashboard in standalone mode.
> - Use the Aspire dashboard with a Python application.

## Prerequisites

To complete this tutorial, you need the following:

- [Docker](https://docs.docker.com/get-docker/)
  - You can use an alternative container runtime, but the commands in this article are for Docker.

> - [Python 3.12 or higher](https://www.python.org/downloads/) locally installed

Reviewer: 3.9 should be fine unless we're doing anything really exotic.

Author: ```suggestion
- [Python 3.9 or higher](https://www.python.org/downloads/) locally installed
```

> - A sample allication

Reviewer: ```suggestion
- A sample application
```

### [FastAPI](#tab/fastapi)

1. Go to the application folder:

    ```Console
    cd msdocs-python-fastapi-webapp-quickstart
    ```

1. Create a virtual environment for the app:

    [!INCLUDE [Virtual environment setup](./includes/standalone/virtual-environment-setup.md)]

1. Install the dependencies:

    ```Console
    pip install -r requirements.txt
    ```

1. Run the app:

    ```Console
    uvicorn main:app --reload
    ```

1. Browse to the sample application at `http://localhost:8000` in a web browser.

    :::image type="content" source="./media/standalone/run-fastapi-app-localhost.png" alt-text="Screenshot of the FastAPI app running locally in a browser.":::

---

## Adding OpenTelemetry

To use the Aspire dashboard with your Python application, you need to install the OpenTelemetry SDK and exporter. The OpenTelemetry SDK provides the API for instrumenting your application, and the exporter sends telemetry data to the Aspire dashboard.

1. Install the OpenTelemetry SDK and exporter:

    ```Console
    pip install opentelemetry-api opentelemetry-sdk opentelemetry-exporter-otlp-proto-grpc
    ```

> 

Reviewer: For FastAPI:

opentelemetry-instrumentation-fastapi

```python
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

# Add this line after the app is created and OTEL is setup
FastAPIInstrumentor.instrument_app(app)
```

For Flask:

opentelemetry-instrumentation-flask

```python
from opentelemetry.instrumentation.flask import FlaskInstrumentor

# add this line after configure_grpc_otel() call
FlaskInstrumentor().instrument()
```

Author: I'll add a pivot that adds the framework specific instrumentation code

---

> title: Using Aspire Dashboard with Python Applications

Reviewer 2: ```suggestion
title: Use the .NET Aspire dashboard with Python apps
```

description: How to use the Aspire Dashboard in a Python application.
ms.date: 08/27/2024
ms.topic: tutorial

> ms.author: aapowell

Reviewer 2: As the `ms.author` you'll be tagged for updates, issues, etc. If you're ok with that, I am too. Otherwise, you can omit the author metadata and that responsibility will fall to me. Again, I'm good either way.

Author: I don't mind either way. If you're happy to take ownership after it's contributed then I'm fine, otherwise I can be kept as the owner.

---

> # Tutorial: Using the Aspire Dashboard with Python Applications

Reviewer 2: ```suggestion
# Tutorial: Use the .NET Aspire dashboard with Python apps
```

> The [Aspire dashboard](overview.md) provides a great UI for viewing telemetry, and is available as a standalone container image that can be used with any OpenTelemetry-enabled app. In this article, you'll learn how to:

Reviewer 2: In docs, we need to always say ".NET Aspire" never just "Aspire", per marketing.

```suggestion
The [.NET Aspire dashboard](overview.md) provides a great user experience for viewing telemetry, and is available as a standalone container image that can be used with any OpenTelemetry-enabled app. In this article, you'll learn how to:
```

> [!div class="checklist"]
>

> > - Start the Aspire dashboard in standalone mode.
> > - Use the Aspire dashboard with a Python application.

Reviewer 2: ```suggestion
> - Start the .NET Aspire dashboard in standalone mode.
> - Use the .NET Aspire dashboard with a Python app.
```

## Prerequisites

To complete this tutorial, you need the following:

> - [Docker](https://docs.docker.com/get-docker/)

Reviewer 2: Why not an OCI-compliant container runtime? We usually specify either Docker or Podman.

Author: How would `An OCI-compliant container runtime, such as [Docker](...) or [Podman](...)` read?

Reviewer 2: Or even just:

```suggestion
- [Docker](https://docs.docker.com/get-docker/) or [Podman](https://podman.io/).
```

  - You can use an alternative container runtime, but the commands in this article are for Docker.

> - [Python 3.9 or higher](https://www.python.org/downloads/) locally installed

Reviewer 2: ```suggestion
- [Python 3.9 or higher](https://www.python.org/downloads/) locally installed.
```

> - A sample application

Reviewer 2: ```suggestion
- A sample application.
```

## Sample application

> This quickstart can be completed using either Flask, Django, or FastAPI. A sample application in each framework is provided to help you follow along with this quickstart. Download or clone the sample application to your local workstation.

Reviewer 2: Now we're a quickstart? We're either a tutorial or a quickstart, not both 😄. Since we have prerequisites listed, that usually implies a tutorial.

```suggestion
This tutorial can be completed using either Flask, Django, or FastAPI. A sample application in each framework is provided to help you follow along with this tutorial. Download or clone the sample application to your local workstation.
```

### [Flask](#tab/flask)

> ```Console

Reviewer 2: ```suggestion
```console
```

git clone https://github.com/Azure-Samples/msdocs-python-flask-webapp-quickstart
```

### [FastAPI](#tab/fastapi)

> ```Console

Reviewer 2: ```suggestion
```console
```

git clone https://github.com/Azure-Samples/msdocs-python-fastapi-webapp-quickstart.git
```

---

To run the application locally:

### [Flask](#tab/flask)

1. Go to the application folder:

>     ```Console

Reviewer 2: ```suggestion
    ```console
```

    cd msdocs-python-flask-webapp-quickstart
    ```

1. Create a virtual environment for the app:

    [!INCLUDE [Virtual environment setup](./includes/standalone/virtual-environment-setup.md)]

1. Install the dependencies:

>     ```Console

Reviewer 2: ```suggestion
    ```python
```

Author: `pip` is a CLI tool (like NuGet), which is why it's `Console` (or should be `console`).

```suggestion
    ```console
```

    pip install -r requirements.txt
    ```

1. Run the app:

>     ```Console

Reviewer 2: ```suggestion
    ```console
```

    flask run
    ```

1. Browse to the sample application at `http://localhost:5000` in a web browser.

    :::image type="content" source="./media/standalone/run-flask-app-localhost.png" alt-text="Screenshot of the Flask app running locally in a browser":::

### [FastAPI](#tab/fastapi)

1. Go to the application folder:

>     ```Console

Reviewer 2: ```suggestion
    ```console
```

    cd msdocs-python-fastapi-webapp-quickstart
    ```

1. Create a virtual environment for the app:

    [!INCLUDE [Virtual environment setup](./includes/standalone/virtual-environment-setup.md)]

1. Install the dependencies:

>     ```Console

Reviewer 2: ```suggestion
    ```python
```

Author: `pip` is a CLI tool (like NuGet), which is why it's `Console` (or should be `console`).

```suggestion
    ```console
```

    pip install -r requirements.txt
    ```

1. Run the app:

>     ```Console

Reviewer 2: ```suggestion
    ```console
```

    uvicorn main:app --reload
    ```

1. Browse to the sample application at `http://localhost:8000` in a web browser.

    :::image type="content" source="./media/standalone/run-fastapi-app-localhost.png" alt-text="Screenshot of the FastAPI app running locally in a browser.":::

---

## Adding OpenTelemetry

> To use the Aspire dashboard with your Python application, you need to install the OpenTelemetry SDK and exporter. The OpenTelemetry SDK provides the API for instrumenting your application, and the exporter sends telemetry data to the Aspire dashboard.

Reviewer 2: ```suggestion
To use the .NET Aspire dashboard with your Python app, you need to install the OpenTelemetry SDK and exporter. The OpenTelemetry SDK provides the API for instrumenting your application, and the exporter sends telemetry data to the .NET Aspire dashboard.
```

1. Install the OpenTelemetry SDK and exporter:

>     ```Console

Reviewer 2: ```suggestion
    ```python
```

Author: `pip` is a CLI tool (like NuGet), which is why it's `Console` (or should be `console`).

```suggestion
    ```console
```

    pip install opentelemetry-api opentelemetry-sdk opentelemetry-exporter-otlp-proto-grpc
    ```

1. Add a new file to your application called `otlp_tracing.py` and add the following code:

>     ```Python

Reviewer 2: ```suggestion
    ```python
```

        handler = LoggingHandler(level=logging.NOTSET, logger_provider=logger_provider)
        handler.setFormatter(logging.Formatter("Python: %(message)s"))

        # Attach OTLP handler to root logger
        logging.getLogger().addHandler(handler)

        tracer = trace.get_tracer(__name__)
        return tracer
    ```

1. Update your application (`app.py` for Flask, `main.py` for FastAPI) to include the imports and call the `configure_oltp_grpc_tracing` function:

    ```python
    import logging
    from otlp_tracing import configure_otel_otlp

    logging.basicConfig(level=logging.INFO)
    tracer = configure_otel_otlp()
    logger = logging.getLogger(__name__)
    ```

1. Replace the `print` calls with `logger.info` calls in your application.
1. Restart your application.

### Framework Specific Instrumentation

This instrumentation has only focused on adding OpenTelemetry to our code. For more detailed instrumentation, you can use the OpenTelemetry Instrumentation packages for the specific frameworks that you are using.

### [Flask](#tab/flask)

1. Install the Flask instrumentation package:

>     ```Console

Reviewer 2: ```suggestion
    ```python
```

Author: `pip` is a CLI tool (like NuGet), which is why it's `Console` (or should be `console`).

```suggestion
    ```console
```

    pip install opentelemetry-instrumentation-flask
    ```

1. Add the following code to your application:

>     ```Python

Reviewer 2: ```suggestion
    ```python
```

    from opentelemetry.instrumentation.flask import FlaskInstrumentor

    # add this line after configure_otel_otlp() call
    FlaskInstrumentor().instrument()
    ```

### [FastAPI](#tab/fastapi)

1. Install the FastAPI instrumentation package:

>     ```Console

Reviewer 2: ```suggestion
    ```python
```

Author: `pip` is a CLI tool (like NuGet), which is why it's `Console` (or should be `console`).

```suggestion
    ```console
```

    pip install opentelemetry-instrumentation-fastapi
    ```

1. Add the following code to your application:

>     ```Python

Reviewer 2: ```suggestion
    ```python
```

    from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

    # add this line after configure_otel_otlp() call
    FastAPIInstrumentor.instrument_app(app)
    ```

---

## Start the Aspire dashboard

To start the Aspire dashboard in standalone mode, run the following Docker command:

```bash
docker run --rm -it -p 18888:18888 -p 4317:18889 --name aspire-dashboard \
    mcr.microsoft.com/dotnet/aspire-dashboard:8.0.0
```

In the Docker logs, the endpoint and key for the dashboard are displayed. Copy the key and navigate to `http://localhost:18888` in a web browser. Enter the key to log in to the dashboard.

> ## Viewing Structured Logs

Reviewer 2: ```suggestion
## View Structured Logs
```

Navigate around the Python application, and you'll see structured logs in the Aspire dashboard. The structured logs page displays logs from your application, and you can filter and search the logs.

:::image type="content" source="./media/standalone/aspire-dashboard-python-logs.png" alt-text="Screenshot of the Aspire dashboard showing the Python application logs":::

## Next steps

> Congratulations! You have successfully used the Aspire dashboard with a Python application. To learn more about the Aspire dashboard, see the [Aspire dashboard overview](overview.md) and how to orchestrate a Python application with the [Aspire App Host](../../get-started/build-aspire-apps-with-python.md).

Reviewer 2: ```suggestion
You have successfully used the .NET Aspire dashboard with a Python application. To learn more about the .NET Aspire dashboard, see the [Aspire dashboard overview](overview.md) and how to orchestrate a Python application with the [.NET Aspire app host](../../get-started/build-aspire-apps-with-python.md).
```
